MapboxNavigation/NavigationMapView.swift

```diff
@@ -487,126 +477,202 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
 
-    func mainRouteCasingStyleLayer(identifier: String, source: MGLSource) -> MGLLineStyleLayer {
-        let mainRouteCasingLayer = MGLLineStyleLayer(identifier: identifier, source: source)
+    @discardableResult func addMainRouteCasingLayer(_ style: MGLStyle, source: MGLSource, lineGradient: NSExpression, below layer: MGLStyleLayer) -> MGLStyleLayer {
+        let customMainRouteCasingLayer = navigationMapViewDelegate?.navigationMapView(self,
+                                                                                      mainRouteCasingStyleLayerWithIdentifier: StyleLayerIdentifier.mainRouteCasing,
+                                                                                      source: source)
+        let currentMainRouteCasingLayer = style.layer(withIdentifier: StyleLayerIdentifier.mainRouteCasing)
+        
+        if let mainRouteCasingLayer = customMainRouteCasingLayer, let _ = currentMainRouteCasingLayer {
+            return mainRouteCasingLayer
+        }
+        
+        if let mainRouteCasingLayer = customMainRouteCasingLayer, currentMainRouteCasingLayer == nil {
+            style.insertLayer(mainRouteCasingLayer, below: layer)
+            return mainRouteCasingLayer
+        }
+        
+        if let mainRouteCasingLayer = currentMainRouteCasingLayer as? MGLLineStyleLayer {
+            return mainRouteCasingLayer
+        }
+        
+        let mainRouteCasingLayer = MGLLineStyleLayer(identifier: StyleLayerIdentifier.mainRouteCasing, source: source)
         mainRouteCasingLayer.predicate = NSPredicate(format: "isAlternateRoute == false")
-        // Default color if no traffic is enabled
         mainRouteCasingLayer.lineColor = NSExpression(forConstantValue: routeCasingColor)
         mainRouteCasingLayer.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 1.5))
         mainRouteCasingLayer.lineJoin = NSExpression(forConstantValue: "round")
         mainRouteCasingLayer.lineCap = NSExpression(forConstantValue: "round")
-
-        if !routeGradientStops.casing.isEmpty {
-            mainRouteCasingLayer.lineGradient = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($lineProgress, 'linear', nil, %@)", NSDictionary(dictionary: routeGradientStops.casing))
-        }
-
+        mainRouteCasingLayer.lineGradient = lineGradient
+        
+        style.insertLayer(mainRouteCasingLayer, below: layer)
+        
         return mainRouteCasingLayer
     }
-
-    func alternativeRouteStyleLayer(identifier: String, source: MGLSource) -> MGLLineStyleLayer {
-        let alternateRoutesLayer = MGLLineStyleLayer(identifier: identifier, source: source)
+    
+    @discardableResult func addAlternateRoutesLayer(_ style: MGLStyle, source: MGLSource, below layer: MGLStyleLayer) -> MGLStyleLayer {
+        let customAlternateRoutesLayer = navigationMapViewDelegate?.navigationMapView(self,
```

Comment:
At this point with use `NavigationMapViewDelegate.navigationMapView(_:alternativeRouteStyleLayerWithIdentifier:source:)` and `NavigationMapViewDelegate.navigationMapView(_:alternateRouteCasingStyleLayerWithIdentifier:source:)`. @1ec5, maybe it'd be better to use the same convention here? As I can see `alternate` is mostly used in `NavigationMapView`, maybe we should use the same naming in public API method?